File: src/main/java/net/elysium/mod/fabricord/Fabricord.java
package net.elysium.mod.fabricord;

import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
import net.elysium.mod.fabricord.discord.DiscordBot;
import net.minecraft.server.MinecraftServer;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class Fabricord implements ModInitializer {

	public static final String MOD_ID = "fabricord";
	static final Logger LOGGER = LogManager.getLogger(Fabricord.class);

	private DiscordBot discordBot;
	private static final ConfigManager configManager = new ConfigManager();

	public Fabricord() {
	}

	@Override
	public void onInitialize() {
		try {
			// Register server lifecycle listeners
			ServerLifecycleEvents.SERVER_STARTING.register(this::onServerStarting);
			ServerLifecycleEvents.SERVER_STOPPING.register(this::onServerStopping);
		} catch (Exception e) {
			LOGGER.error("Error during initialization: ", e);
		}
	}

	private void onServerStarting(MinecraftServer server) {
		try {
			this.discordBot = new DiscordBot(server);
			configManager.checkAndCreateConfig();
			discordBot.registerEventListeners();
			discordBot.startBot();
		} catch (Exception e) {
			LOGGER.error("Error during server starting: ", e);
		}
	}

	private void onServerStopping(MinecraftServer server) {
		try {
			discordBot.stopBot();
			this.discordBot = null;
		} catch (Exception e) {
			LOGGER.error("Error during server stopping: ", e);
		}
	}
}

File: src/main/kotlin/com/inf_ruxy/fabricord/Fabricord.kt
package com.inf_ruxy.fabricord

import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

/**
 * This class represents the main mod initializer for the Fabricord mod.
 * It implements the ModInitializer interface, which means it provides an onInitialize method that gets called when the mod is initialized.
 */
@Suppress("unused")
class Fabricord : ModInitializer {

    companion object {
        /**
         * The MOD_ID constant holds the ID of the mod, which is "fabricord".
         */
        const val MOD_ID = "fabricord"

        /**
         * The logger is used to log information, warnings, errors etc. during the mod's lifecycle.
         * It is named after the mod ("Fabricord").
         */
        val logger: Logger = LogManager.getLogger("Fabricord")
    }

    /**
     * This method is called when the mod is initialized.
     * It logs the start of the initialization, registers server lifecycle events, and logs the end of the initialization.
     */
    override fun onInitialize() {
        // Log the start of the initialization
        logger.info("    ")
        logger.info("Initializing $MOD_ID...")
        logger.info("    ")

        // Register server lifecycle events
        ServerLifecycleEvents.SERVER_STARTING.register(FabricordApi::serverStarting)
        ServerLifecycleEvents.SERVER_STARTED.register(FabricordApi::serverStarted)
        ServerLifecycleEvents.SERVER_STOPPING.register(FabricordApi::serverStopping)
    }

}